src/multiversioning.zig

```diff
@@ -0,0 +1,836 @@
+        versions: [past_releases_max]u32 = std.mem.zeroes([past_releases_max]u32),
+        checksums: [past_releases_max]u128 = std.mem.zeroes([past_releases_max]u128),
+
+        /// Offsets are relative to the start of the pack (`.tbmvp`) offset.
+        offsets: [past_releases_max]u32 = std.mem.zeroes([past_releases_max]u32),
+
+        sizes: [past_releases_max]u32 = std.mem.zeroes([past_releases_max]u32),
+
+        /// Normally version upgrades are allowed to skip to the latest. If a corresponding release
+        /// is set to true here, it must be visited on the way to the newest version.
+        visits: [past_releases_max]bool = std.mem.zeroes([past_releases_max]bool),
+
+        pub fn init(count: u32, past_versions: struct {
+            versions: []const u32,
+            checksums: []const u128,
+            offsets: []const u32,
+            sizes: []const u32,
+            visits: []const bool,
+        }) PastVersionPack {
+            assert(past_versions.versions.len == count);
+            assert(past_versions.checksums.len == count);
+            assert(past_versions.offsets.len == count);
+            assert(past_versions.sizes.len == count);
+            assert(past_versions.visits.len == count);
+            assert(count <= past_releases_max);
+
+            var pack = PastVersionPack{};
+            pack.count = count;
+
+            std.mem.copy(u32, &pack.versions, past_versions.versions);
+            std.mem.copy(u128, &pack.checksums, past_versions.checksums);
+            std.mem.copy(u32, &pack.offsets, past_versions.offsets);
+            std.mem.copy(u32, &pack.sizes, past_versions.sizes);
+            std.mem.copy(bool, &pack.visits, past_versions.visits);
+
+            return pack;
+        }
+    };
+
+    /// The current version is executed differently to past versions embedded in the pack, so store
+    /// it separately. See exec_latest vs exec_release.
+    current_version: u32,
+
+    /// The AEGIS128L checksum of the direct output of `zig build`, for the current_version, before
+    /// any objcopy or build magic has been performed.
+    /// Used when extracting binaries from past releases to ensure a hash chain - this will be put
+    /// in the entry in past.checksums for this release by the build script.
+    current_checksum: u128,
+
+    /// The AEGIS128L checksum of the binary, if the `.tbmvm` section were zeroed out. Used to
+    /// validate that the binary itself is not corrupt. Putting this in requires a bit of trickery:
+    /// - inject a zero `.tbmvm` section of the correct size,
+    /// - compute the hash,
+    /// - update the section with the correct data.
+    /// Used to ensure we don't try and exec into a corrupt binary.
+    checksum_without_metadata: u128 = 0,
+
+    past: PastVersionPack = .{},
+    reserved: [16384]u8 = std.mem.zeroes([16384]u8),
+
+    /// Covers MultiversionMetadata[0..@sizeOf(MultiversionMetadata) - @sizeOf(u128)].
+    checksum_metadata: u128 = undefined,
```

Comment:
Hmm - I don't think so, since I'm not sure there's a usecase for having the binary itself encrypted in that manner. Then again, I've seen weirder things being required for compliance in the past, so I'm not opposed to putting it in pre-emptively :stuck_out_tongue:.

Reply:
I don't think it is necessary either, just didn't want it to get skipped if I was missing something.